<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2015 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

*This code is used with permission by Google Genomics*
*https://github.com/googlegenomics*

# Sample-Level QC

```{r echo=FALSE, eval=FALSE}
######################[ CHANGE ME ]##################################
# This codelab assumes that the current working directory is where the Rmd file resides.
setwd("/Users/gmcinnes/GitHub/bigquery-genomics-qc/Visualization")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "gbsc-gcp-project-mvp"
#####################################################################
```

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
# Set up for BigQuery access.
source("./rHelpers/setup.R")
```


```{r}
tableReplacement <- list("_THE_TABLE_"="va_aaa_pilot_data.all_genomes_gvcfs",
                          "_THE_EXPANDED_TABLE_"="va_aaa_pilot_data.all_genomes_expanded_vcfs_java",
                          "_PATIENT_INFO_"="va_aaa_pilot_data.patient_info")
sampleData <- read.csv("/Users/gmcinnes/GitHub/mvp_aaa_codelabs/qc/data/patient_info.csv")
sampleInfo <- select(sampleData, sample_id=Catalog.ID, gender=Gender)

#ibs <- read.table("./data/platinum-genomes-ibs.tsv",
#                  col.names=c("sample1", "sample2", "ibsScore", "similar", "observed"))

# To run this against other public data, source in one of the dataset helpers.  For example:
# source("./rHelpers/pgpCGIOnlyDataset.R")
```


* [Missingness Rate](#missingness-rate)
* [Singleton Rate](#singleton-rate)
* [Heterozygosity Rate and Inbreeding Coefficient](#homozygosity-rate-and-inbreeding-coefficient)
* [Sex Inference](#sex-inference)

## Missingness Rate

Identify all the genomes that have a sample level missingness greater than 0.1.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/missingness-sample-level-fail.sql",
                                  project=project,
                                  replacements=tableReplacement)
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

And visualizing the results:
```{r sampleMissingnessFail, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(result) +
  geom_point(aes(x=sample_id, y=missingness)) +
  theme(axis.text.x=if(nrow(result) <= 20)
    {element_text(angle = 90, hjust = 1)} else {element_blank()}) +
  xlab("Sample") +
  ylab("Missingness Rate") +
  ggtitle("Genome-Specific Missingness")
```

## Singleton Rate

Identify genomes that have a singleton rate more than 3 standard deviations from the mean.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/private-variants-metrics.sql",
                                  project=project,
                                  replacements=tableReplacement)
```

Displaying the results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

Determine cutoffs
```{r}
mean = result$average
stddev = result$stddev
max_value = mean + 3*stddev
min_value = mean - 3*stddev
```

Determine which genomes are outside our desired range

```{r message=FALSE, warning=FALSE, comment=NA}
values = list("_MAX_VALUE_" = max_value,
              "_MIN_VALUE_" = min_value)
result <- DisplayAndDispatchQuery("../sql/private-variants-fail.sql",
                                  project=project,
                                  replacements=c(tableReplacement, values))
```

Displaying the results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

And visualizing the results:
```{r singletonsFail, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(result) +
  geom_point(aes(x=call_call_set_name, y=private_variant_count)) +
  theme(axis.text.x=if(nrow(result) <= 20)
  {element_text(angle = 90, hjust = 1)} else {element_blank()}) +
  xlab("Sample") +
  ylab("Number of Singletons") +
  ggtitle("Count of Singletons Per Genome")
```

## Homozygosity Rate and Inbreeding Coefficient

Identify genomes that have an inbreeding coefficient outside the desired range.

```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/inbreeding-coefficient-metrics.sql",
                                  project=project,
                                  replacements=tableReplacement)
```

Displaying the results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Determine the cutoffs:
```{r}
mean = result$average
stddev = result$stddev
max_value = mean + 3*stddev
min_value = mean - 3*stddev
```

Determine which genomes are outside our desired range
```{r message=FALSE, warning=FALSE, comment=NA}
values = list("_MAX_VALUE_" = max_value,
              "_MIN_VALUE_" = min_value)
result <- DisplayAndDispatchQuery("../sql/inbreeding-coefficient-fail.sql",
                                  project=project,
                                  replacements=c(tableReplacement, values))
```

Displaying the results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(result), type="html", include.rownames=F)
```

And visualizing the results:
```{r homozygosityFail, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(result) +
  geom_text(aes(x=O_HOM, y=E_HOM, label=call_call_set_name), hjust=-1, vjust=0) +
  xlab("Observed Homozygous Variants") +
  ylab("Expected Homozygous Variants") +
  ggtitle("Homozygosity")
```

## Sex Inference

For each genome, compare the gender from the sample information to the heterozygosity rate on the chromosome X calls.
```{r message=FALSE, warning=FALSE, comment=NA}
result <- DisplayAndDispatchQuery("../sql/gender-check-fail.sql",
                                  project=project,
                                  replacements=tableReplacement)
```
Number of rows returned by this query: `r nrow(result)`.

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Let's join this with the sample information:
```{r message=FALSE, warning=FALSE, comment=NA}
joinedResult <- inner_join(result, sampleInfo)
```

And visualize the results:
```{r genderFail, fig.align="center", fig.width=10, message=FALSE, comment=NA}
ggplot(joinedResult) +
  geom_point(aes(x=sample_id, y=perct_het_alt_in_snvs, color=gender)) +
  theme(axis.text.x=if(nrow(result) <= 20)
    {element_text(angle = 90, hjust = 1)} else {element_blank()}) +
  xlab("Sample") +
  ylab("Heterozygosity Rate ") +
  ggtitle("Heterozygosity Rate on the X Chromosome")

```

