<!-- R Markdown Documentation, DO NOT EDIT THE PLAIN MARKDOWN VERSION OF THIS FILE -->

<!-- Copyright 2015 Google Inc. All rights reserved. -->

<!-- Licensed under the Apache License, Version 2.0 (the "License"); -->
<!-- you may not use this file except in compliance with the License. -->
<!-- You may obtain a copy of the License at -->

<!--     http://www.apache.org/licenses/LICENSE-2.0 -->

<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Variant-Level QC

```{r echo=FALSE, eval=FALSE}
######################[ CHANGE ME ]##################################
# This codelab assumes that the current working directory is where the Rmd file resides.
setwd("/Users/gmcinnes/GitHub/bigquery-genomics-qc/Visualization")

# Set the Google Cloud Platform project id under which these queries will run.
project <- "gbsc-gcp-project-mvp"
#####################################################################
```

```{r echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
# Set up for BigQuery access.
source("./rHelpers/setup.R")
```

* [Missingness Rate](#missingness-rate)
* [Blacklisted Variants](#blacklisted-variants)
* [Heterozygous Haplotype](#heterozygous-haplotype)
* [Ti/Tv by Genomic Window](#titv-by-genomic-window)
* [Ti/Tv by Depth](#titv-by-depth)
* [Ti/Tv by Alternate Allele Counts](#titv-by-alternate-allele-counts)
* [Hardy Weinberg Equilibrium](#hardy-weinberg-equilibrium)

```{r}
queryReplacements <- list("_THE_TABLE_"="va_aaa_pilot_data.all_genomes_gvcfs_20150514",
                          "_THE_EXPANDED_TABLE_"="va_aaa_pilot_data.all_genomes_expanded_vcfs_java3",
                          "_PATIENT_INFO_"="va_aaa_pilot_data.patient_info",
                          "_BLACKLISTED_TABLE_"="resources.blacklisted_positions")
```

## Missingness Rate

Identify all variants with a missingness rate greater than a specified cutoff.

```{r message=FALSE, warning=FALSE, comment=NA}
cutoff = list("_CUTOFF_"="0.1")
sortAndLimit <- list("#_ORDER_BY_" = "LIMIT 1000")
outputTable = 'qc_tables.variant_missingness'
result <- DisplayAndDispatchQuery("../sql/variant-level-missingness-fail.sql",
                                  project=project,
                                  replacements=c(cutoff, queryReplacements),
                                  outputTable=outputTable)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(result)), type="html", include.rownames=F)
}
```

## Blacklisted Variants

```{r message=FALSE, warning=FALSE, comment=NA}
query <- "../sql/blacklisted-variants.sql"
sortAndLimit <- list("#_ORDER_BY_" = "LIMIT 1000")
outputTable = 'qc_tables.blacklisted'
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(queryReplacements),
                                  outputTable=outputTable)
```

Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

First few results
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

## Heterozygous Haplotype
For each variant within the X and Y chromosome, identify heterozygous variants in male genomes.

```{r message=FALSE, warning=FALSE, comment=NA}
sortAndLimit <- list("_LIMIT_" = "LIMIT 1000")
outputTable = 'qc_tables.heterozygous_haplotype'
result <- DisplayAndDispatchQuery("../sql/sex-chromosome-heterozygous-haplotypes.sql",
                                  project=project,
                                  replacements=c(queryReplacements),
                                  outputTable=outputTable)
```
Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

Displaying the first few rows of the dataframe of results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
if(is.null(result)) { 
  cat("**None**") 
} else {
  print(xtable(head(result)), type="html", include.rownames=F)
}
```

## Ti/Tv By Genomic Window
```{r message=FALSE, warning=FALSE, comment=NA}
query <- "../sql/titv-by-genomic-window-fail.sql"
sortAndLimit <- list("#_ORDER_BY_" = "LIMIT 1000")
max <- 3.0
min <- 1.5
cutoffs <- list("_MAX_" = max, "_MIN_" = min)
outputTable = 'qc_tables.titv_genomic_window'
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(queryReplacements, cutoffs),
                                  outputTable=outputTable)
```

Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

First few results
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

## Ti/Tv By Depth

We want to identify all the regions of the genome where the Ti/Tv ratio is outside of the expected range.  Another method we can use to do this is calculating the transition-transversion ratio by depth of coverage.  
```{r message=FALSE, warning=FALSE, comment=NA}
query <- "../sql/titv-by-depth-fail.sql"
max <- 3
min <- 1.5
cutoffs <- list("_MAX_" = max, "_MIN_" = min)
sortAndLimit <- list("#_ORDER_BY_" = "LIMIT 1000")
outputTable = 'qc_tables.titv_depth'
result <- DisplayAndDispatchQuery(query,
                                  project=project,
                                  replacements=c(queryReplacements, cutoffs),
                                  outputTable)
```

Number of rows returned by this query: **`r if(is.null(result)) { "None" } else { nrow(result) }`**.

First few results
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```


## Ti/Tv By Alternate Allele Counts
Collect all the alternate allele counts that are outside our desired range.
```{r message=FALSE, warning=FALSE, comment=NA}
query <- "../sql/titv-by-allternate-allele-fail.sql"
max <- 3
min <- 1.5
cutoffs <- list("_MAX_" = max, "_MIN_" = min)
sortAndLimit <- list("#_ORDER_BY_" = "LIMIT 1000")
#result <- DisplayAndDispatchQuery(query,
#                                  project=project,
#                                  replacements=c(queryReplacements, cutoffs, sortAndLimit))
```

Get the variant ids for all the failed groups


I'll write this query later.  There are no variants that fail this qc step for our current dataset.

## Hardy Weinberg Equilibrium

Here we want to identify the variants that are out of Hardy Weinberg Equilibrium.  We want to remove the top 0.05 quantile of variants, so first we have to define what the cutoff for the chi squared value should be.
```{r message=FALSE, warning=FALSE, comment=NA}
quantile <- list("_QUANTILE_" = 1999) # <- Define quantile by number. 
                                  # The 1999th quantile selects the value that partitions the top 0.05% of values, 
                                  # assuming there are 2000 quantiles.
result <- DisplayAndDispatchQuery("../sql/hwe-quantile.sql",
                                  project=project,
                                  replacements=c(queryReplacements, quantile))
```

Displaying the results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```

Determine the cutoffs:
```{r}
maxChiSq = result$quantile
```
Cutoff: `r maxChiSq`

Determine which genomes are outside our desired range
```{r message=FALSE, warning=FALSE, comment=NA}
values = list("_CUTOFF_" = maxChiSq)
sortAndLimit <- list("#_ORDER_BY_" = "LIMIT 1000")
outputTable = 'qc_tables.hardy_weinberg'
result <- DisplayAndDispatchQuery("../sql/hwe-fail.sql",
                                  project=project,
                                  replacements=c(queryReplacements, values),
                                  outputTable=outputTable)
```

Displaying the first few results:
```{r echo=FALSE, message=FALSE, warning=FALSE, comment=NA, results="asis"}
print(xtable(head(result)), type="html", include.rownames=F)
```







