# Compute private variants counts for each sample.
SELECT 
ROUND(AVG(private_variant_count), 3) AS average,
ROUND(STDDEV(private_variant_count), 3) AS stddev
FROM (
  SELECT
  call.call_set_name,
  COUNT(call.call_set_name) AS private_variant_count,
  FROM (
    SELECT
    reference_name,
    start,
    GROUP_CONCAT(CASE WHEN cnt = 1 THEN 'S'
                 WHEN cnt = 2 THEN 'D'
                 ELSE STRING(cnt) END) AS SINGLETON_DOUBLETON,
    reference_bases,
    alternate_bases,
    GROUP_CONCAT(call.call_set_name) AS call.call_set_name,
    GROUP_CONCAT(genotype) AS genotype,
    SUM(num_samples_with_variant) AS num_samples_with_variant
    FROM (
      SELECT
      reference_name,
      start,
      reference_bases,
      alternate_bases,
      alt_num,
      call.call_set_name,
      GROUP_CONCAT(STRING(call.genotype)) WITHIN call AS genotype,
      SUM(call.genotype == alt_num) WITHIN call AS cnt,
      COUNT(call.call_set_name) WITHIN RECORD AS num_samples_with_variant
      FROM (
        FLATTEN((
          SELECT
          reference_name,
          start,
          reference_bases,
          alternate_bases,
          POSITION(alternate_bases) AS alt_num,
          call.call_set_name,
          call.genotype,
          FROM
          [_THE_TABLE_]
          # Optionally add a clause here to limit the query to a particular
          # region of the genome.
          #_WHERE_
          OMIT call IF EVERY(call.genotype = -1)
        ), alternate_bases)
      )
      OMIT RECORD IF alternate_bases IS NULL
      HAVING
      cnt > 0
    )
    GROUP EACH BY
    reference_name,
    start,
    reference_bases,
    alternate_bases
    HAVING
    num_samples_with_variant = 1
  )
  GROUP BY
  call.call_set_name)

